### PR TITLE
feat: Add blob scalar type

### LINKED ISSUE #2090: Blob scalar type
Adding support for a blob scalar type will simplify storage of binary objects such as images, audio files, etc. in DefraDB.

### PR DESCRIPTION
**Pull request by nasdf**
## Relevant issue(s)

Resolves #2090 

## Description

This PR adds a blob scalar type to the schema system.

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

Added integration tests.

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

File: request/graphql/schema/descriptions.go
Comment (nasdf) [reply request/graphql/schema/descriptions.go:17]: I'm open to that. Should we move the other types there as well?

Comment (AndrewSisley) [reply request/graphql/schema/descriptions.go:17]: This makes me nervous, I think we added DateTime to that package, but `graphql-go` is supposed to be a general purpose gql package, not a defra package.  And we may/do want to get rid of it/replace it at somepoint

Comment (jsimnz) [reply request/graphql/schema/descriptions.go:17]: > I think we added DateTime to that package

No, DateTime was already defined as a scalar in the `graphql-go` package, but there were some problems with it as it relates to Defra integration, there was PR that got merged on the `graphql-go` package related to the DateTime stuff, but it wasn't the implementation/definition of the DateTime scalar. (reference: https://github.com/sourcenetwork/defradb/pull/931 and https://github.com/sourcenetwork/graphql-go/pull/8)

I do agree though that we shouldn't have to define the scalar in the `graphql-go` package, it exposes the necessary types/functions for us to define our own scalars outside that package, exactly as they're being used here. Moreover, the fact that we have our own defined custom scalars implies that w.e package we may replace the `graphql-go` package with, should implement the "base" scalars as defined in the GQL spec, and all we have to worry about is the specific "custom" scalars that are neatly organized in a single place.

File: request/graphql/schema/descriptions.go
Comment (jsimnz) [inline request/graphql/schema/descriptions.go:35]: suggestion: Related to the other comment from Fred, I think it should be made clear that these are custom scalars, as the comment on line 36 suggests ("More custom ones to come"). Just a quick comment separating base scalars from custom scalars for clarity 👍

File: client/descriptions.go
Comment (AndrewSisley) [inline client/descriptions.go:207]: todo: I think this is a strange way of representing this, and that `[Byte]` is more sensible and fits much better with the rest of the types, and any potential future `Byte`, `[Byte!]` or `Byte!` types.

File: request/graphql/schema/types/scalars_test.go
Comment (AndrewSisley) [inline request/graphql/schema/types/scalars_test.go:43]: question: Why is this invalid?

todo: Please document why this is invalid, if it remains invalid.

File: request/graphql/schema/types/scalars.go
Comment (AndrewSisley) [inline request/graphql/schema/types/scalars.go:60]: todo: Similar to another comment, please document this - it looks like a silent error, but I only know that by guessing given the:
```
if err != nil {
	return nil
}
```
stuff in other code blocks

File: tests/integration/schema/updates/add/field/kind/invalid_test.go
Comment (AndrewSisley) [inline tests/integration/schema/updates/add/field/kind/invalid_test.go:67]: todo: Please add a tests in `.../schema/updates/add/field/kind/foo.go` that tests that this field works with PatchSchema

File: request/graphql/schema/types/scalars.go
Comment (AndrewSisley) [inline request/graphql/schema/types/scalars.go:22]: question:  If it is an array of bytes, why does it only appear to accept string values?

todo: I'm not sure if it does only accept string values, could you add an integration test documenting the beheviour when provided an array of numbers:
```
testUtils.CreateDoc{
	Doc: `{
		"name": "John",
		"data": [0,1,2,3...]
	}`,
},
```

File: request/graphql/schema/types/scalars.go
Comment (AndrewSisley) [inline request/graphql/schema/types/scalars.go:30]: todo: Please document this `return nil`

**Review by AndrewSisley — CHANGES_REQUESTED** [review]
I have a handful of localized questions/todos that I would like resolving before merging, sorry about the hassle :)

File: request/graphql/schema/types/scalars.go
Comment (AndrewSisley) [inline request/graphql/schema/types/scalars.go:26]: question: Why are you encoding/decoding at all here (including in ParseValue)? 

todo: The `graphql` library lacks any documentation here, but we can add some to our code to document why `Serialize` and `ParseValue` exist.

File: client/descriptions.go
Comment (nasdf) [reply client/descriptions.go]: Updated to `Blob` as we discussed in the standup.

File: request/graphql/schema/types/scalars.go
Comment (nasdf) [reply request/graphql/schema/types/scalars.go:48]: I'm just following how the other types work in the `graphql-go` library. It seems like `nil` is the return value when the type cannot be parsed.

File: request/graphql/schema/types/scalars.go
Comment (nasdf) [reply request/graphql/schema/types/scalars.go]: I believe this is a bug in the`client.Document` API. You can also do this to a string field:

```
testUtils.CreateDoc{
		Doc: `{
			"name":  12345
		}`,
}
testUtils.CreateDoc{
		Doc: `{
			"name":  false
		}`,
}
```

Comment (AndrewSisley) [reply request/graphql/schema/types/scalars.go]: Not quite a bug, we just dont do any checking of the saved value.  The read should fail though I think.  Does the read of `[0,1,2,3...]` fail?  Can we please document the behaviour, as it is a more likely question than assigning `1234` to a boolean :)

**Review by AndrewSisley — APPROVED** [review]
Looks good to me, thanks for this, and the extra tweaks Keenan - I'm approving now with one open/pending discussion RE why you are converting to hex and back that should be resolved before merge :)

**Comment by nasdf** [comment]
> Looks good to me, thanks for this, and the extra tweaks Keenan - I'm approving now with one open/pending discussion RE why you are converting to hex and back that should be resolved before merge :)

Thanks! I refactored the parsing logic so that the value will always be a hex string.

File: request/graphql/schema/types/scalars.go
Comment (fredcarle) [inline request/graphql/schema/types/scalars.go:55]: nitpick: "to a hex string"?

**Review by islamaliev — COMMENTED** [review]
Looks good, but one test is missing.

**Review by islamaliev — APPROVED** [review]
good job, Keenan!

**Pull request merged**
